src/main/java/com/rtb/projectmanagementtool/project/ProjectData.java

```diff
@@ -66,38 +50,43 @@ public ProjectData(String name, String description, long creatorId) {
    * @param entity the datastore entity to parse object from
    */
   public ProjectData(Entity entity) {
-    datastore = DatastoreServiceFactory.getDatastoreService();
-
     this.id = (Long) entity.getKey().getId();
     this.name = (String) entity.getProperty(PROPERTY_NAME);
     this.description = (String) entity.getProperty(PROPERTY_DESCRIPTION);
 
-    Gson gson = new Gson();
-
-    // sets the type of containers to store parsed JSON in
-    Type mapType = new TypeToken<HashMap<Long, User>>() {}.getType();
-    Type setType = new TypeToken<HashSet<Long>>() {}.getType();
+    Object userIdProperty;
+    if ((userIdProperty = entity.getProperty(PROPERTY_TASK_IDS)) != null) {
+      this.taskIds = new HashSet<Long>((ArrayList<Long>) userIdProperty);
+    } else {
+      this.taskIds = new HashSet<Long>();
+    }
 
-    this.userIds = gson.fromJson((String) entity.getProperty(PROPERTY_USER_IDS), mapType);
-    this.taskIds = gson.fromJson((String) entity.getProperty(PROPERTY_TASK_IDS), setType);
+    this.userIds = new HashMap<Long, User>();
+    // can store map in datastore entities using EmbeddedEntity
+    EmbeddedEntity ee = (EmbeddedEntity) entity.getProperty(PROPERTY_USER_IDS);
+    if (ee != null) {
+      for (String key : ee.getProperties().keySet()) {
+        this.userIds.put(Long.parseLong(key), User.valueOf((String) ee.getProperty(key)));
+      }
+    }
   }
 
-  /** Saves the object as a datastore entity. */
-  private void toEntity() {
-    Gson gson = new Gson();
-    try {
-      Key entityKey = KeyFactory.createKey("Project", this.id);
-      Entity entity = datastore.get(entityKey);
-
-      entity.setProperty(PROPERTY_NAME, this.name);
-      entity.setProperty(PROPERTY_DESCRIPTION, this.description);
-      entity.setProperty(PROPERTY_USER_IDS, gson.toJson(this.userIds));
-      entity.setProperty(PROPERTY_TASK_IDS, gson.toJson(this.taskIds));
-
-      datastore.put(entity);
-    } catch (Exception exception) {
-      System.out.println("Error: Project not found.");
+  /**
+  * @return the entity representation of this class 
+  */
+  public Entity toEntity() {
+    Entity entity = new Entity("Project");
+    entity.setProperty(PROPERTY_NAME, this.name);
+    entity.setProperty(PROPERTY_DESCRIPTION, this.description);
+    entity.setProperty(PROPERTY_TASK_IDS, this.taskIds);
+
+    EmbeddedEntity userIdsEntity = new EmbeddedEntity();
+    for (Long key : this.userIds.keySet()) {
+      userIdsEntity.setProperty(key.toString(), this.userIds.get(key).name());
     }
+    entity.setProperty(PROPERTY_USER_IDS, userIdsEntity);
```

Comment:
I wonder if there's another way to store a map in datastore w/o having to be concerned with parsing it when retrieving. It's straightforward with a set as seen in like 81, but not so much with map. EmbeddedEntity is the best way I found

Reply:
You can store a json string representing the whole map.  Or a Repeated property with json strings per item.  But either way you'd still need to parse (maybe json has an advantage because of built in parsing/encoding methods).  Storing them as an EmbeddedEntity I think is ok as well.